src/commonMain/kotlin/com/charleskorn/kaml/Yaml.kt

```diff
@@ -18,20 +18,109 @@
 
 package com.charleskorn.kaml
 
+import com.charleskorn.kaml.internal.bufferedSource
 import kotlinx.serialization.DeserializationStrategy
+import kotlinx.serialization.SerializationStrategy
 import kotlinx.serialization.StringFormat
 import kotlinx.serialization.modules.EmptySerializersModule
 import kotlinx.serialization.modules.SerializersModule
+import okio.Buffer
+import okio.Sink
+import okio.Source
+import okio.use
+import org.snakeyaml.engine.v2.api.StreamDataWriter
 
-public expect class Yaml(
-    serializersModule: SerializersModule = EmptySerializersModule(),
-    configuration: YamlConfiguration = YamlConfiguration(),
+public class Yaml(
+    override val serializersModule: SerializersModule = EmptySerializersModule(),
+    public val configuration: YamlConfiguration = YamlConfiguration(),
 ) : StringFormat {
-    public val configuration: YamlConfiguration
-
-    public fun <T> decodeFromYamlNode(deserializer: DeserializationStrategy<T>, node: YamlNode): T
 
     public companion object {
-        public val default: Yaml
+        public val default: Yaml = Yaml()
+    }
+
+    public fun <T> decodeFromYamlNode(
+        deserializer: DeserializationStrategy<T>,
+        node: YamlNode,
+    ): T {
+        val input = YamlInput.createFor(node, this, serializersModule, configuration, deserializer.descriptor)
+        return input.decodeSerializableValue(deserializer)
+    }
+
+    override fun <T> decodeFromString(
+        deserializer: DeserializationStrategy<T>,
+        string: String,
+    ): T {
+        return decodeFromSource(deserializer, string.bufferedSource())
+    }
+
+    public fun <T> decodeFromSource(
+        deserializer: DeserializationStrategy<T>,
+        source: Source,
+    ): T {
+        val rootNode = parseToYamlNodeFromSource(source)
+
+        val input = YamlInput.createFor(rootNode, this, serializersModule, configuration, deserializer.descriptor)
+        return input.decodeSerializableValue(deserializer)
+    }
+
+    public fun parseToYamlNode(string: String): YamlNode =
+        parseToYamlNodeFromSource(string.bufferedSource())
+
+    internal fun parseToYamlNodeFromSource(source: Source): YamlNode {
+        val parser = YamlParser(source)
+        val reader =
+            YamlNodeReader(parser, configuration.extensionDefinitionPrefix, configuration.allowAnchorsAndAliases)
+        val node = reader.read()
+        parser.ensureEndOfStreamReached()
+        return node
+    }
+
+    public fun <T> encodeToSink(
+        serializer: SerializationStrategy<T>,
+        value: T,
+        sink: Sink,
+    ) {
+        encodeToBufferDataWriter(serializer, value).use { writer ->
+            writer.readAll(sink)
```

Comment:
In the case of writing to a `Sink`, couldn't we write directly to the sink rather than through a buffer? We'd need another implementation of `StreamDataWriter` that writes directly to a `Sink`

Reply:
Alternatively, `BufferedDataWriter` could take a `Sink` (and be renamed to `SinkDataWriter`) and then we can pass the `Sink` in this method and a `Buffer` in `encodeToString`

Reply:
Good idea, done!

I had to use a BufferedSink instead, because there's no `Sink.writeUtf8()`.